design/external-references.md

```diff
@@ -0,0 +1,344 @@
+
+**CloudFormation imports/exports as a serialization context**: on top of that, we will implement a serialization/deserialization context based on AWS CloudFormation's import/exports mechanism. "writing" a value means "output/export" and "reading" a value means "import". The "export name" will be used as a prefix that represent an object, and sub-objects will be serialized by adding another component to the export name.
+
+**Convenience methods**: now that constructs can be serializable through imports/exports, we can implement a set of convenience methods for each AWS resource to provide nice ergonomics for the specific use case of exporting and importing resources across apps.
+
+## Design
+
+### Serialization
+
+If a construct is said to be serializable, it must implement the `ISerializable` interface:
+
+```ts
+interface ISerializable {
+  serialize(ctx: ISerializationContext): void;
+}
+```
+
+When an object is serialized, the `serialize` method is called with an object which implements the following interface:
+
+```ts
+interface ISerializationContext {
+  writeString(key: string, value: string, options?: SerializationOptions): void
+  writeObject(key: string, obj: ISerializable, options?: SerializationOptions): void;
+}
+
+interface SerializationOptions {
+  description?: string;
+}
+```
+
+The serialization context allows the object serialize itself through key/value strings via calls to `writeString(key, value)`. If the object encapsulates another serializable object, it can use `writeObject(key, obj)`, which will result in a subsequent call to the sub-object's `serialize` method with an appropriate context.
+
+### Deserialization
+
+To support deserialization, classes must also include a public static `deserializeXxx` method which reads the object from a deserialization context and returns an object that implements the resource type interface:
+
+```ts
+class MyResource extends Construct implements ISerializable {
+  static deserializeMyResource(ctx: IDeserializationContext): IMyResource;
+}
+```
+
+The deserialization context is an object that implements the following interface:
+
+```ts
+interface IDeserializationContext {
+  scope: Construct;
+  id: string;
+  readString(key: string): string;
+  readObject(key: string): IDeserializationContext;
+}
+```
+
+`readString(key)` can be used to read values stored by `writeString`. and `readObject(key)` returns a deserialization context for composite deserialization written via `writeObject`.
+
+Since `deserializeXxx` will normally need to create new construct objects, the deserialization context should supply a consistent `scope` and `id` which can be used to instantiate a construct object that represents this object.
+
+Implementers of `deserializeXxx` should check if a construct with `id` already exists within `scope` and return it instead of instantiating and new object. This will satisfy REQ6.
```

Comment:
Only if the scope is the same. REQ6 made it sound as if it would be the same object in the entire app.

Reply:
It should be the same object for the entire stack and it is the deserializatiom context’s job to ensure by always passing in the same scope and id. See the code below for an example how they are used.

I am not super fond of that model, but couldn’t come up with something more elegant. Ideas are welcome.

Reply:
@rix0rrr does that make sense?

Reply:
Yes, sure. In my head the subobjects would be created under the deserialized object, but that's (1) impossible and also (2) unnecessary, so why even bother :).

I think this is a good solution.